realtek: Improve SMI polling

oliv3r opened the pull request:
Currently, we have to put fake SMI bus numbers on SFP links, which use in-band SMI messaging. Lets improve our routine a little bit so we can drop fakes.

oliv3r committed:
realtek: eth: mdio: Improve SMI polling configuration based on .dts

Improves error detection of the association of smi addresses to ports
on RTL93xx. Allows empty smi addresses for ports using a PHY linked to an
internal SerDes.

While here, ensure our variable is large enough to properly hold the
result of `of_property_read_u32`.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek: dts: Drop needless SMI assignment on XGS1250

The SFP port doesn't have an MDIO connection, as the MDIO is in-band
with the serdes signal. To statisfy the driver probing, we falsely
assigned MDI bus 4, which does not even exist.

With the latest fixes in the driver, we no longer need to supply this
fake MDI bus.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r commented:
@svanheule had this one in the queue for a while, and should have it to avoid having fake-values in dts files. I see people are getting confused by this and making up values, which doesn't make sense of course.

svanheule commented on target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c:
```diff
@@ -2173,33 +2175,46 @@ static int rtl838x_mdio_init(struct rtl838x_eth_priv *priv)
+		if (of_property_read_u32_array(dn, "rtl9300,smi-address", &smi_addr[0], 2)) {
+			/* Integrated PHYs associated to a SerDes do not have an smi_bus */
+			if (priv->phy_is_internal[pn] && priv->sds_id[pn] >= 0) {
+				priv->smi_bus[pn] = -1;
```
Setting a u32 to -1, are we? ;)

In any case, I don't think you should even try reading this property if you know you're not going to use it for the current port.

oliv3r replied:
hah, fair point; but, I think this is still a legal (C) construct. You are basically setting it to u32_max; but I think I saw that macro somewhere, so we can use that macro instead.

And that's exactly what this code does, if we don't have the property, we need to set the marker, as the `smi_bus` property is being used all over the code.

I don't intend to 'change the world' with this change, and it's birgers change of course, as he wrote all of this anyway.

svanheule replied:
The problem with Birger's patches was, and still is, code quality. I'm fine with the rationale of this patch set, but the implementation needs to improve before I'm merging this.

oliv3r replied:
That's a bit over the top yeah? This all will be thrown away, this is not usable in its current form, it needs to be completely rewritten. We lack full understanding of a lot of things to be able to fully re-write it. The code quality of this patch is more then adequate, considering what's underneath. Gold-plating a turd won't help anyone.

Obvious mistakes, absolutely. But we are swimming in a pile of diarrhea, and you want to spend enormous amounts of time, on something that will never end up upstream (because its just wrong in many ways, wrong location, wrong subsystem, wrong 'type' of driver). I don't know how much spare time you have for all of this, but the rest of us _are_ limited herein.

If this where something that would possibly go upstream, sure, we can do better.

But the purpose of this, is to gain users, fix bugs, get more functionality and learn how the hardware works. This allows us to write actual upstream able code.

What we currently have upstream, has been problematic already (bugs in gpio for example) due to us not understanding the hardware. The drivers upstream even need to be improved a little. The SPI driver for example, while basic and functional, lacks a lot still. And that is a 200 line driver.

With this attitude, we won't ever finish support for these switches honestly. And we seriously lack manpower (and understanding of the hardware) to even consider going the 'pure mainline route'. Ideal, yes, realistic ... you guys have been at this for 3 years. Is there anybody contributing or USING mainline?

Babysteps are needed really. Otherwise, we just end up alienating developers (which has been happening all over, and yes, that code quality was poor, yet we merged the clock, timer etc drivers, which also are far from ideal).

In short; we have what we have. It's ugly, its bad. But don't let perfect be the enemy of better. Small betters is what we need for now. We'd love to make everything great again. Baby. Steps.

svanheule commented on target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c:
```diff
@@ -190,7 +190,7 @@ struct rtl838x_eth_priv {
 	u16 rxrings;
 	u16 rxringlen;
-	u8 smi_bus[MAX_PORTS];
+	u32 smi_bus[MAX_PORTS];
```
Since it's not in the commit message: why is the width change needed?

oliv3r replied:
Idk; birger did this, but I'm pretty sure, because he's using `of_property_read_u32`.

svanheule replied:
Fair enough, although the property is read into an intermediate variable and should be bound checked (it isn't) before it is assigned to the `priv::smi_bus` and `priv::smi_addr` properties, which both fit into a u8.

oliv3r replied:
true, but lets just keep it at u32; no needs for checking; as we do the bounds checking elsewhere anyway (SMI_MAX_BUSSES). Also, this is all 'throw-away-code' anyway :)

svanheule commented on target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c:
```diff
@@ -2030,6 +2030,8 @@ static int rtl931x_mdio_reset(struct mii_bus *bus)
 		u32 pos;
 
+		if ((priv->smi_bus[i] < 0) || (priv->smi_bus[i] > MAX_SMI_BUSSES))
+			continue;
```
Same problem with the first condition.

oliv3r replied:
Actually, this is a valid point, as we don't have to check for anything less then 0 :) so I can remove that first bit. as it was always a no-op; it won't hurt, but its just stupid indeed.

svanheule replied:
`MAX_SMI_BUSSES` is also confusing to me. Is it the highest valid SMI bus index (`MAX_SMI_BUS`)? Or is it the number of valid SMI bus indices (`SMI_BUS_COUNT`)? If it's the latter, the second comparison is also wrong and should be "greater than or equal".

svanheule replied:
Going from the message on the second patch, it appears 4 is the number of busses, as an index of 4 is invalid. So this also needs to be fixed.

oliv3r replied:
what do you mean with 'also'?

svanheule replied:
The "greater than MAX_SMI_BUSSES" comparison.

svanheule commented on target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c:
```diff
@@ -1922,7 +1922,7 @@ static int rtl930x_mdio_reset(struct mii_bus *bus)
 		int pos;
 
-		if (priv->smi_bus[i] > 3)
+		if ((priv->smi_bus[i] < 0) || (priv->smi_bus[i] > MAX_SMI_BUSSES))
```
The first condition is pointless. u32-s can't be smaller than 0.

svanheule reviewed (commented):
Didn't dive into the related code, so mainly some comments about obvious issues.

svanheule commented on target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c:
```diff
@@ -1922,7 +1923,7 @@ static int rtl930x_mdio_reset(struct mii_bus *bus)
 		int pos;
 
-		if (priv->smi_bus[i] > 3)
+		if (priv->smi_bus[i] > MAX_SMI_BUSSES)
```
This also changes the check from "> 3> to "> 4".

oliv3r replied:
Well spotted

```suggestion
		if (priv->smi_bus[i] >= MAX_SMI_BUSSES)
```

oliv3r commented on target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c:
```diff
@@ -2030,6 +2031,8 @@ static int rtl931x_mdio_reset(struct mii_bus *bus)
 	for (int i = 0; i < 56; i++) {
 		u32 pos;
 
+		if (priv->smi_bus[i] > MAX_SMI_BUSSES)
```
```suggestion
		if (priv->smi_bus[i] >= MAX_SMI_BUSSES)
```

oliv3r committed:
realtek: eth: mdio: Avoid -1 marker

The sds_id variable is an unsigned. While it is perfectly legal to store
a `-1` to indicate a special state, it is the same as using `U32_MAX`
which is more correct. The value is the same, but the reader will
immediately know it is not a signed int.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek: eth: mdio: Avoid using magic values for rtl930x

We know what these numbers mean, so lets avoid being magical about them.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r commented:
I stopped carrying about getting this in early. Who needs it anyway.
